### PR TITLE
backend/cos: Support endpoint

### PR DESCRIPTION
**Pull request by lyu571**
Support endpoint for the COS backend

## Target Release
1.5.x

### ENHANCEMENTS

-  Provide an endpoint way to access backend of Tencentcloud COS. This method can provide support for internal network URLs for customers.

### CODE REVIEW

**Comment by lyu571** [comment]
Hi @crw @likexian plz help to review my PR.

**Comment by likexian** [comment]
LGTM

**Comment by lyu571** [comment]
HI @crw  plz help to review my PR. :)

File: internal/backend/remote-state/cos/backend.go [hunk 1]
```diff
@@ -87,6 +92,18 @@ func New() backend.Backend {
 				Required:    true,
 				Description: "The name of the COS bucket",
 			},
+			"endpoint": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "The custom endpoint for the COS API. eg: http://cos-internal.{Region}.tencentcos.cn, both http or https can be accepted.",
```
Comment (kmoe) [inline internal/backend/remote-state/cos/backend.go:98]:
```suggestion
				Description: "The custom endpoint for the COS API, e.g. http://cos-internal.{Region}.tencentcos.cn. Both HTTP and HTTPS are accepted.",
```

File: internal/backend/remote-state/cos/backend.go [hunk 2]
```diff
@@ -87,6 +92,18 @@ func New() backend.Backend {
...
+				DefaultFunc: schema.EnvDefaultFunc(PROVIDER_ENDPOINT, nil),
+			},
+			"domain": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc(PROVIDER_DOMAIN, nil),
+				Description: "The root domain of the API request, Default is `tencentcloudapi.com`.",
```
Comment (kmoe) [inline internal/backend/remote-state/cos/backend.go:105]:
```suggestion
				Description: "The root domain of the API request. Default is tencentcloudapi.com.",
```

File: internal/backend/remote-state/cos/backend.go [hunk 3]
```diff
@@ -231,6 +248,33 @@ func (b *Backend) configure(ctx context.Context) error {
 		return err
 	}
 
+	if v, ok := data.GetOk("domain"); ok {
+		b.domain = v.(string)
+		log.Printf("[DEBUG][backend] set domain for TencentCloud api client. domain:[%s]", b.domain)
```
Comment (kmoe) [inline internal/backend/remote-state/cos/backend.go:253]:
```suggestion
		log.Printf("[DEBUG] Backend: set domain for TencentCloud API client. Domain: [%s]", b.domain)
```

File: internal/backend/remote-state/cos/backend.go [hunk 4]
```diff
@@ -231,6 +248,33 @@ func (b *Backend) configure(ctx context.Context) error {
...
+	}
+	// set url as endpoint when it provided
```
Comment (kmoe) [inline internal/backend/remote-state/cos/backend.go:255]:
```suggestion
	// set url as endpoint when provided
```

File: internal/backend/remote-state/cos/backend.go [hunk 5]
```diff
@@ -231,6 +248,33 @@ func (b *Backend) configure(ctx context.Context) error {
...
+			return fmt.Errorf("[backend] invalid URL, must be: %v", "http(s)://cos-internal.{Region}.tencentcos.cn")
+		}
+
+		protocol := matches[1]
+		region := matches[3]
+
+		// URL after converting
+		newUrl := fmt.Sprintf("%s://%s.cos-internal.%s.tencentcos.cn", protocol, b.bucket, region)
+		u, err = url.Parse(newUrl)
+		log.Printf("[DEBUG][backend] set cos url as:[%s]", newUrl)
```
Comment (kmoe) [inline internal/backend/remote-state/cos/backend.go:272]:
```suggestion
		log.Printf("[DEBUG] Backend: set COS URL as: [%s]", newUrl)
```

File: internal/backend/remote-state/cos/backend_test.go [hunk 6]
```diff
@@ -201,6 +222,44 @@ func TestBackendWithEncryption(t *testing.T) {
...
+	backend.TestBackendStateLocks(t, be0, be1)
+	backend.TestBackendStateForceUnlock(t, be0, be1)
+}
+
+func setupBackendWithEndpoint(t *testing.T, bucket, prefix, key string, encrypt bool) backend.Backend {
+	t.Helper()
+
+	skip := os.Getenv("TF_COS_APPID") == ""
+	if skip {
+		t.Skip("This test require setting TF_COS_APPID environment variables")
```
Comment (kmoe) [inline internal/backend/remote-state/cos/backend_test.go:246]:
```suggestion
		t.Skip("This test requires setting the TF_COS_APPID environment variable")
```

File: website/docs/language/settings/backends/cos.mdx [hunk 7]
```diff
@@ -64,6 +64,8 @@ The following configuration options or environment variables are supported:
 - `encrypt` - (Optional) Whether to enable server side encryption of the state file. If it is true, COS will use 'AES256' encryption algorithm to encrypt state file.
 - `acl` - (Optional) Object ACL to be applied to the state file, allows `private` and `public-read`. Defaults to `private`.
 - `accelerate` - (Optional) Whether to enable global Acceleration. Defaults to `false`.
+- `endpoint` - (Optional) The Custom Endpoint for the COS backend. It supports environment variables `TENCENTCLOUD_ENDPOINT`.
+- `domain` - (Optional) The root domain of the API request. Defaults to `tencentcloudapi.com`. It supports environment variables `TENCENTCLOUD_DOMAIN`.
```
Comment (kmoe) [inline website/docs/language/settings/backends/cos.mdx:68]:
```suggestion
- `domain` - (Optional) The root domain of the API request. Defaults to `tencentcloudapi.com`. It supports the environment variable `TENCENTCLOUD_DOMAIN`.
```

File: website/docs/language/settings/backends/cos.mdx [hunk 8]
```diff
@@ -106,4 +108,22 @@ $ export TENCENTCLOUD_ASSUME_ROLE_ARN="qcs::cam::uin/xxx:roleName/yyy"
 $ export TENCENTCLOUD_ASSUME_ROLE_SESSION_NAME="my-session-name"
 $ export TENCENTCLOUD_ASSUME_ROLE_SESSION_DURATION=3600
 $ terraform plan
+```
+
+### Endpoint
+If provided with an endpoint url, Terraform will attempt to access COS backend by the `endpoint` configuration or environment variables `TENCENTCLOUD_ENDPOINT`.
```
Comment (kmoe) [inline website/docs/language/settings/backends/cos.mdx:114]:
```suggestion
If provided with an endpoint URL, Terraform will attempt to access the COS backend by the `endpoint` configuration or the environment variable `TENCENTCLOUD_ENDPOINT`.
```

File: website/docs/language/settings/backends/cos.mdx [hunk 9]
```diff
@@ -106,4 +108,22 @@ $ export TENCENTCLOUD_ASSUME_ROLE_ARN="qcs::cam::uin/xxx:roleName/yyy"
...
+
+A typical endpoint like this: `http://cos-internal.{Region}.tencentcos.cn`, both http or https can be accepted.
```
Comment (kmoe) [inline website/docs/language/settings/backends/cos.mdx:116]:
```suggestion
A typical endpoint looks like this: `http://cos-internal.{Region}.tencentcos.cn`. Both HTTP and HTTPS are accepted.
```

**Review by kmoe — CHANGES_REQUESTED** [review]
Some docs comments

**Comment by crw** [comment]
Hi @lyu571, as long as you are OK accepting this docs changes, we can merge this PR. Thanks!

**Comment by lyu571** [comment]
> Hi @lyu571, as long as you are OK accepting this docs changes, we can merge this PR. Thanks!

Hi @crw I've accepted all docs comments. Plz continues the processing. Thank you.

**Review by kmoe — APPROVED** [review]

**Comment by lyu571** [comment]
Hi @kmoe and @crw, the feature `endpoint` was required by our customer who already waited a while. Seems the latest(1.5.7) Terraform release doesn't include this PR. Could you put the release plan for this PR with high priority? 

Maybe change the milestone from 1.6.0 to 1.5.x?

**Comment by crw** [comment]
Hi @lyu571, the first version of the product with this feature will be `1.6.0`. You should be able to see it in `1.6.0-beta1`. Thanks!

**Comment by lyu571** [comment]
> Hi @lyu571, the first version of the product with this feature will be `1.6.0`. You should be able to see it in `1.6.0-beta1`. Thanks!

Oh, I see.
Looks like this version is a pre-release version that our customer cannot use in an official way.
May I ask what time the official `1.6.0` be released/available?

**Comment by crw** [comment]
Early October. `1.6.0-beta2` was released today, then there will be two release candidate releases.

**Comment by lyu571** [comment]
> Early October. `1.6.0-beta2` was released today, then there will be two release candidate releases.

ok. Thanks!

**Pull request merged**